httpcache.go

```diff
@@ -498,6 +516,39 @@ func headerAllCommaSepValues(headers http.Header, name string) []string {
 	return vals
 }
 
+// cachingReadCloser is a wrapper around ReadCloser R that calls OnEOF
+// handler with a full copy of the content read from R when EOF is
+// reached.
+type cachingReadCloser struct {
+	// Underlying ReadCloser.
+	R io.ReadCloser
+	// OnEOF is called with a copy of the content of R when EOF is reached.
+	OnEOF func(io.Reader)
+
+	buf bytes.Buffer // buf stores a copy of the content of R.
+}
+
+// Read reads the next len(p) bytes from R or until R is drained. The
+// return value n is the number of bytes read. If R has no data to
+// return, err is io.EOF and OnEOF is called with a full copy of what
+// has been read so far.
+func (r *cachingReadCloser) Read(p []byte) (n int, err error) {
+	n, err = r.R.Read(p)
+	if n > 0 {
+		if _, err = r.buf.Write(p[:n]); err != nil {
```

Comment:
Shadowing can be very dangerous.

This line just introduced a very serious but subtle issue.

It's unfortunate to see that none of our tests caught this.

In my previous suggestion, I was careful to make it a new `err` value:

```Go
if _, err := r.buf.Write(p[:n]); err != nil {
```

However, you've changed it to be a simple assignment instead:

```
if _, err = r.buf.Write(p[:n]); err != nil {
```

That means the original `err` value returned from `r.R.Read` is being overwritten here.

So, if `n > 0` if block is entered, by the time it returns, `err` will always be `nil`, even if the original read error was `io.EOF` or something else non-`nil`.

In my personal Go style guide, I only use `err` name if its value is consumed either on next line, or the one after. Otherwise, I try to give it a longer name, to prevent this kind of accidental shadowing from happening (and to improve readability). So let's do that here:

```Go
n, readError := r.R.Read(p)
// ... adjust the rest of the code accordingly.
```